File: go/models/audit_log.go
// Owner: backend (audit logging)
// Consumers: all services (audit trail)

package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// AuditActionType represents the type of action in the audit log
type AuditActionType string

const (
	// Mission actions
	AuditActionMissionCreated   AuditActionType = "mission_created"
	AuditActionMissionClaimed   AuditActionType = "mission_claimed"
	AuditActionMissionCompleted AuditActionType = "mission_completed"
	AuditActionMissionArchived  AuditActionType = "mission_archived"
	AuditActionMissionDeleted   AuditActionType = "mission_deleted"

	// Asset actions
	AuditActionAssetStatusChanged AuditActionType = "asset_status_changed"

	// Event actions (created by correlation service, read by backend)
	AuditActionEventCorrelated AuditActionType = "event_correlated"
	AuditActionEventSuggested  AuditActionType = "event_suggested"
	AuditActionEventApproved   AuditActionType = "event_approved"
	AuditActionEventRejected   AuditActionType = "event_rejected"

	// Operator actions
	AuditActionOperatorOrder AuditActionType = "operator_order"
	AuditActionOperatorNote  AuditActionType = "operator_note"

	// Operational command actions (from AI Command Center)
	AuditActionCommandReceived AuditActionType = "command_received"
	AuditActionCommandAccepted AuditActionType = "command_accepted"
	AuditActionCommandDeclined AuditActionType = "command_declined"
	AuditActionCommandStarted  AuditActionType = "command_started"
	AuditActionCommandCompleted AuditActionType = "command_completed"
)

// AuditLog represents a complete audit trail entry for mission actions
type AuditLog struct {
	ID        primitive.ObjectID `json:"id" bson:"_id,omitempty"`
	MissionID string             `json:"missionId" bson:"mission_id"`
	Timestamp time.Time          `json:"timestamp" bson:"timestamp"`

	// Action details
	ActionType AuditActionType `json:"actionType" bson:"action_type"`
	ActorType  string          `json:"actorType" bson:"actor_type"` // "operator", "asset", "ai_agent", "system"
	ActorID    string          `json:"actorId" bson:"actor_id"`
	ActorName  string          `json:"actorName,omitempty" bson:"actor_name,omitempty"`

	// Target of action
	TargetType string `json:"targetType,omitempty" bson:"target_type,omitempty"` // "asset", "dispatch", "event"
	TargetID   string `json:"targetId,omitempty" bson:"target_id,omitempty"`
	TargetName string `json:"targetName,omitempty" bson:"target_name,omitempty"`

	// Details
	Action    string                 `json:"action" bson:"action"` // Human-readable action
	Details   map[string]interface{} `json:"details" bson:"details"`
	CreatedAt time.Time              `json:"createdAt" bson:"created_at"`
}

File: typescript/src/models/AuditLog.ts
/**
 * AuditLog model
 * Converted from go/models/audit_log.go
 * Owner: backend (audit logging)
 * Consumers: all services (audit trail)
 */

/** AuditActionType represents the type of action in the audit log */
export const AuditActionType = {
  // Mission actions
  MISSION_CREATED: 'mission_created' as const,
  MISSION_CLAIMED: 'mission_claimed' as const,
  MISSION_COMPLETED: 'mission_completed' as const,
  MISSION_ARCHIVED: 'mission_archived' as const,
  MISSION_DELETED: 'mission_deleted' as const,

  // Asset actions
  ASSET_STATUS_CHANGED: 'asset_status_changed' as const,

  // Event actions (created by correlation service, read by backend)
  EVENT_CORRELATED: 'event_correlated' as const,
  EVENT_SUGGESTED: 'event_suggested' as const,
  EVENT_APPROVED: 'event_approved' as const,
  EVENT_REJECTED: 'event_rejected' as const,

  // Operator actions
  OPERATOR_ORDER: 'operator_order' as const,
  OPERATOR_NOTE: 'operator_note' as const,

  // Operational command actions (from AI Command Center)
  COMMAND_RECEIVED: 'command_received' as const,
  COMMAND_ACCEPTED: 'command_accepted' as const,
  COMMAND_DECLINED: 'command_declined' as const,
  COMMAND_STARTED: 'command_started' as const,
  COMMAND_COMPLETED: 'command_completed' as const,
};

export type AuditActionTypeValue = typeof AuditActionType[keyof typeof AuditActionType];

/** AuditLog represents a complete audit trail entry for mission actions */
export interface AuditLog {
  id: string;
  missionId: string;
  timestamp: string; // ISO 8601

  // Action details
  actionType: AuditActionTypeValue;
  actorType: string; // "operator", "asset", "ai_agent", "system"
  actorId: string;
  actorName?: string;

  // Target of action
  targetType?: string; // "asset", "dispatch", "event"
  targetId?: string;
  targetName?: string;

  // Details
  action: string; // Human-readable action
  details: Record<string, any>;
  createdAt: string; // ISO 8601
}
